File: src/dp/5-longest-palindromic-substring.cpp
class Solution {
public:
    bool IsPalindromic(const string& s, int i, int j)
    {
        if (i == j)
            return true;
        int len = j - i + 1;
        int mid = (i + j) / 2;
        int left = (len & 1) ? mid - 1 : mid;
        int right = mid + 1;

        while (left >= i && right <= j) {
            if (s[left] != s[right])
                return false;
        }
        return true;
    }

    string longestPalindrome(string s)
    {
        if (s.size() <= 1)
            return s;

        int beg = 0, maxLen = 1;
        bool dp[s.size()][s.size()];
        for (int i = 0; i < s.size(); ++i) {
            dp[i][i] = true;
        }

        for (int len = 2; len <= s.size(); ++len) {
            for (int i = 0; i < s.size(); ++i) {
                int j = len + i - 1;
                if (j >= s.size())
                    break;
                if (s[i] != s[j])
                    dp[i][j] = false;
                else {
                    if (j - i < 3)
                        dp[i][j] = true;
                    else
                        dp[i][j] = dp[i + 1][j - 1];
                }
                if (dp[i][j] && j - i + 1 > maxLen) {
                    maxLen = j - i + 1;
                    beg = i;
                }
            }
        }
        return s.substr(beg, maxLen);
    }
};

File: go/top100/5-longest-palindromic-substring.go
package main

import "fmt"

// 动态规划: dp[i][j]表示i-j范围是否是回文串
// 1. i == j -> dp[i][j] = true
// 2. i == j - 1 -> dp[i][j] = s[i] == s[j]
// 3. dp[i][j] = dp[i+1][j-1] & s[i] == s[j]

func longestPalindrome(s string) string {
	dp := make([][]bool, len(s))
	beg, maxLen := 0, 1
	for i := 0; i < len(s); i++ {
		dp[i] = make([]bool, len(s))
		dp[i][i] = true
	}
	for l := 2; l <= len(s); l++ {
		for i := 0; i < len(s); i++ {
			j := l + i - 1
			if j >= len(s) {
				break
			}
			if s[i] != s[j] {
				dp[i][j] = false
			} else {
				if j-i < 3 {
					dp[i][j] = true
				} else {
					dp[i][j] = dp[i+1][j-1]
				}
			}
			if dp[i][j] && j-i+1 > maxLen {
				beg = i
				maxLen = j - i + 1
			}
		}
	}
	return s[beg : beg+maxLen]
}
func main() {
	fmt.Println(longestPalindrome("babab"))
}
